ring/replication_set_tracker.go

```diff
@@ -465,3 +466,74 @@ func (t *zoneAwareContextTracker) cancelAllContexts(cause error) {
 		delete(t.cancelFuncs, instance)
 	}
 }
+
+type inflightInstanceTracker struct {
+	mx       sync.Mutex
+	inflight []map[*InstanceDesc]struct{}
+
+	// expectMoreInstances is true if more instances are expected to be added to the tracker.
+	expectMoreInstances bool
+}
+
+func newInflightInstanceTracker(sets []ReplicationSet) *inflightInstanceTracker {
+	// Init the inflight tracker.
+	inflight := make([]map[*InstanceDesc]struct{}, len(sets))
+	for idx, set := range sets {
+		inflight[idx] = make(map[*InstanceDesc]struct{}, len(set.Instances))
+	}
+
+	return &inflightInstanceTracker{
+		inflight:            inflight,
+		expectMoreInstances: true,
+	}
+}
+
+// addInstance adds the instance for replicationSetIdx to the tracker.
+//
+// addInstance is idempotent.
+func (t *inflightInstanceTracker) addInstance(replicationSetIdx int, instance *InstanceDesc) {
+	t.mx.Lock()
```

Comment:
At first I was panicking but then I removed it. Reason is that I looked at `DoUntilQuorumWithoutSuccessfulContextCancellation()`. Code is tricky, but I think there could be some races in case of errors where the callback function may called after `DoUntilQuorumWithoutSuccessfulContextCancellation()` returns with a canceled context, so I've removed the panic.

For our case, that we care is that we'll wait for all instances added before `DoUntilQuorumWithoutSuccessfulContextCancellation()` returns. If a race adds an instance slightly after it will still be handled correctly, but there's no correctness issue if won't be waited for.